aptos-move/block-executor/src/lib.rs

```diff
@@ -147,3 +149,56 @@ mod txn_last_input_output;
 #[cfg(test)]
 mod unit_tests;
 pub mod view;
+
+/// In standard BlockSTM, the internal states assume contiguous transactions indices.
+/// To also support sharded execution (where each shard gets a subsequence of the transactions
+/// and possibly waits for transaction results from other shards),
+/// an index mapping is need as an additional input to our existing BlockSTM implementation.
+#[derive(Clone)]
+pub struct IndexMapping {
+    pub indices: Vec<TxnIndex>,
+    /// A TxnIndex -> local position mapping.
+    /// Currently implemented as a `Vec` of size equal to the block size, assuming it's not too large.
+    pub inverses: Vec<usize>,
+    /// A fake index representing the end of the transaction list.
+    pub end_index: TxnIndex,
+}
+
+impl IndexMapping {
+    pub fn new(indices: Vec<TxnIndex>, block_size: usize) -> Self {
+        let mut inverses = vec![usize::MAX; block_size];
+        for (pos, &index) in indices.iter().enumerate() {
+            inverses[index as usize] = pos;
+        }
+        Self {
+            indices,
+            inverses,
+            end_index: TxnIndex::MAX,
```

Comment:
Why MAX?

Reply:
Magic number is hacky, but it minimizes the change. Doing the option thing has the cost of changing `execution_idx: AtomicU32` into `execution_idx: Atomic<Option<U32>>` which means we have to write some additional if/else I assume.

Reply:
I'd prefer to use the block size instead of MAX